File: frontend/src/app/features/game/models/game-events.ts
export type MovementDirection = 'up' | 'down' | 'left' | 'right';
export type TankStatus = 'active' | 'destroyed';
export type ConnectionStatus = 'disconnected' | 'connecting' | 'connected' | 'reconnecting';
export type RoomStatus = 'waiting' | 'inProgress' | 'finished';

export interface PlayerState {
  playerId: string;
  playerName: string;
  tankType: string;
  x: number;
  y: number;
  health: number;
  score: number;
  ammunition: number;
  kills: number;
  deaths: number;
  direction: MovementDirection;
  status: TankStatus;
  isConnected: boolean;
  isHost: boolean;
  lastDirection?: MovementDirection;
}

export type TankState = PlayerState;

export interface PlayerJoinedEvent extends PlayerState {}

export interface PlayerMovementEvent {
  playerId: string;
  playerName: string;
  direction: MovementDirection;
  x: number;
  y: number;
}

export interface ChatMessageEvent {
  playerId: string;
  playerName: string;
  message: string;
  sentAt: string;
}

export interface BulletState {
  id: string;
  ownerId: string;
  x: number;
  y: number;
  direction: MovementDirection;
}

export interface BlockDestroyedEvent {
  playerId: string;
  row: number;
  column: number;
}

export interface RoomSummary {
  roomCode: string;
  hostPlayerId: string;
  hostPlayerName: string;
  playerCount: number;
  maxPlayers: number;
  mapName: string;
  isPrivate: boolean;
  status: RoomStatus;
}

export interface GameStateSnapshot {
  roomCode: string;
  matchId: string | null;
  status: RoomStatus;
  hostPlayerId: string;
  maxPlayers: number;
  mapName: string;
  players: PlayerState[];
  bullets: BulletState[];
  destroyedBlocks: BlockDestroyedEvent[];
  serverTime: string;
}

export interface MatchStartedEvent {
  matchId: string;
  startedAt: string;
}

export interface MatchEndedEvent {
  matchId: string;
  winnerId: string;
  winnerName: string;
  finishedAt: string;
}

export interface CollisionEvent {
  playerId: string;
  playerName: string;
  targetId: string;
  targetName: string;
  damage: number;
  sentAt: string;
}

export interface GameOverEvent {
  playerId: string;
  playerName: string;
  targetId: string;
  targetName: string;
  sentAt: string;
}

File: backend/api/Hubs/DTOs/GameEvents.cs
namespace backend.api.Hubs.DTOs;

public enum RoomStatus
{
    Waiting,
    InProgress,
    Finished
}

public sealed record CreateRoomRequest(
    int MaxPlayers,
    string MapName,
    string? Password,
    string? Region = null
);

public sealed record JoinRoomRequest(
    string RoomCode,
    string? Password
);

public sealed record RoomSummary(
    string RoomCode,
    Guid HostPlayerId,
    string HostPlayerName,
    int PlayerCount,
    int MaxPlayers,
    string MapName,
    bool IsPrivate,
    RoomStatus Status
);

public sealed record MovementInput(string Direction, long Sequence = 0);

public sealed record ChatMessageInput(string Message);

public sealed record ChatMessageEvent(
    Guid PlayerId,
    string PlayerName,
    string Message,
    DateTime SentAt
);

public sealed record PlayerStateDto(
    Guid PlayerId,
    string PlayerName,
    string TankType,
    int X,
    int Y,
    int Health,
    int Score,
    int Ammunition,
    int Kills,
    int Deaths,
    string Direction,
    string Status,
    bool IsConnected,
    bool IsHost
);

public sealed record BulletStateDto(
    Guid Id,
    Guid OwnerId,
    double X,
    double Y,
    string Direction
);

public sealed record BlockDestroyedEvent(
    Guid PlayerId,
    int Row,
    int Column
);

public sealed record PlayerMovementEvent(
    Guid PlayerId,
    string PlayerName,
    string Direction,
    int X,
    int Y
);

public sealed record PlayerHitEvent(
    Guid AttackerId,
    Guid TargetId,
    int Damage,
    int RemainingHealth,
    DateTime ServerTime
);

public sealed record PlayerEliminatedEvent(
    Guid PlayerId,
    Guid? EliminatedBy,
    int Placement,
    DateTime ServerTime
);

public sealed record MatchStartedEvent(Guid MatchId, DateTime StartedAt);

public sealed record MatchEndedEvent(
    Guid MatchId,
    Guid WinnerId,
    string WinnerName,
    DateTime FinishedAt
);

public sealed record GameStateSnapshot(
    string RoomCode,
    Guid? MatchId,
    RoomStatus Status,
    Guid HostPlayerId,
    int MaxPlayers,
    string MapName,
    IReadOnlyCollection<PlayerStateDto> Players,
    IReadOnlyCollection<BulletStateDto> Bullets,
    IReadOnlyCollection<BlockDestroyedEvent> DestroyedBlocks,
    DateTime ServerTime
);
